gulp-tasks/build-packages.js

```diff
@@ -25,39 +30,63 @@ const ERROR_NO_NAMSPACE = oneLine`
   JavaScript. Please fix for:
 `;
 
-/*
- * This function takes an object like { nested: { foo: bar } } and a string
- * like `nested.foo` and returns true if that value is defined on the obejct.
- */
-const testObjectValueExists = (object, nestedPath) => {
-  const pieces = nestedPath.split('.');
-  let currentRoot = object;
-  for (const piece of pieces) {
-    if (!currentRoot[piece]) {
-      return false;
-    }
-    currentRoot = currentRoot[piece];
+// This makes Rollup assume workbox-* will be added to the global
+// scope and replace references with the core namespace
+const globals = (moduleId) => {
+  // This regex matches for (workbox-*)(/any/path/here.mjs)
+  const workboxModuleIdRegex = /(workbox-\w*)([/\w.]*)*/g;
+  const result = workboxModuleIdRegex.exec(moduleId);
+  if (!result) {
+    throw new Error(`Unknown global module ID: ${moduleId}`);
+  }
+
+  const packageName = result[1];
+  let importFilePath = null;
+  if (result.length === 3) {
+    importFilePath = result[2];
+  }
+
+  if (importFilePath) {
+    throw new Error(oneLine`
+      All imports of workbox-* modules must be done from the top level export.
+      (i.e. import * from 'workbox-*') This ensures that the browser
+      namespacing works correctly. Please remove '${importFilePath}' from
+      the import '${moduleId}'.
+    `);
+  }
+
+  // Get a package's browserNamespace so we know where it will be
+  // on the global scope (i.e. workbox.????)
+  let browserNamespace = null;
+  const packagePath = path.join(__dirname, '..', 'packages', packageName);
+  try {
+    const pkg = require(path.join(packagePath, 'package.json'));
+    browserNamespace = pkg.workbox.browserNamespace;
```

Comment:
Nit: you could make this `const browserNamespace = ...` and then move

```js
return `${constants.NAMESPACE_PREFIX}.${browserNamespace}`;
```

to the next line, and not have to declare `let browserNamespace = null;` outside of the `try{}` block.